### PR TITLE
Low-level client support for Polls

### PR DESCRIPTION
**Pull request by martinmitrevski**
### 🔗 Issue Links

Resolves https://github.com/GetStream/ios-issues-tracking/issues/795.

### 🎯 Goal

Add support in the LLC for polls.

### 📝 Summary

- Exposed a `PollController` for dealing with a poll
- Exposed a `PollVoteListController` for dealing with vote querying
- Exposed new models and events

### 🛠 Implementation

Similar to how we do things in other places.

### 🎨 Showcase

_Add relevant screenshots and/or videos/gifs to easily see what this PR changes, if applicable._

| Before | After |
| ------ | ----- |
|  img   |  img  |

### 🧪 Manual Testing Notes

The UI implementation is part of the `polls` branch in the SwiftUI SDK. It should be used for testing there.

### ☑️ Contributor Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

Alt Text

### CODE REVIEW

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    ///
+    /// - Parameters:
+    ///   - text: The text of the suggested poll option.
+    ///   - position: An optional position for the suggested option.
+    ///   - custom: An optional dictionary of custom data for the suggested option.
+    ///   - completion: A closure to be called upon completion, with an optional `Error` if something went wrong.
+    public func suggestPollOption(
+        text: String,
+        position: Int? = nil,
+        custom: [String: RawJSON]? = nil,
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollController.swift:188]: We use `userData` quite a bit for cases like this. Would it make sense to unify the naming here?

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    internal var multicastDelegate: MulticastDelegate<PollControllerDelegate> = .init() {
+        didSet {
+            stateMulticastDelegate.set(mainDelegate: multicastDelegate.mainDelegate)
+            stateMulticastDelegate.set(additionalDelegates: multicastDelegate.additionalDelegates)
+        }
+    }
+    
+    let ownVotesQuery: PollVoteListQuery
+    
+    private var pollObserver: EntityDatabaseObserverWrapper<Poll, PollDTO>?
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollController.swift:57]: Is there any specific reason why both observers are not created with `lazy var` properties?

File: Sources/StreamChat/Controllers/PollController/PollVoteListController.swift [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        }
+        _basePublishers = BasePublishers(controller: self)
+        return _basePublishers as? BasePublishers ?? .init(controller: self)
+    }
+    
+    private let eventsController: EventsController
+    private let pollsRepository: PollsRepository
+    private let environment: Environment
+
+    /// Creates a new `UserListController`.
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollVoteListController.swift:109]: > `PollVoteListController`

File: Sources/StreamChat/Database/DTOs/PollDTO.swift [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    @NSManaged var pollDescription: String?
+    @NSManaged var enforceUniqueVote: Bool
+    @NSManaged var id: String
+    @NSManaged var name: String
+    @NSManaged var updatedAt: DBDate?
+    @NSManaged var voteCount: Int
+    @NSManaged var custom: Data?
+    @NSManaged var voteCountsByOption: [String: Int]?
+    @NSManaged var isClosed: Bool
+    @NSManaged var maxVotesAllowed: NSNumber?
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/PollDTO.swift:23]: Int?

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    }
+    
+    @discardableResult
+    func savePollVote(
+        payload: PollVotePayload,
+        query: PollVoteListQuery?,
+        cache: PreWarmedCache?
+    ) throws -> PollVoteDTO {
+        guard let poll = try poll(id: payload.pollId) else {
+            throw ClientError.Unexpected()
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/PollVoteDTO.swift:99]: `MessageDoesNotExist` but for polls would be nice since then it prints out the id at least, e.g.`PollDoesNotExist(pollId:)`

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            option: option,
+            user: userDto,
+            context: self,
+            cache: nil
+        )
+        
+        dto.createdAt = Date().bridgeDate
+        dto.updatedAt = Date().bridgeDate
+        dto.pollId = pollId
+        dto.isAnswer = true // TODO: fix
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/PollVoteDTO.swift:164]: Highlighting the `TODO`

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    }
+    
+    func pollVote(id: String, pollId: String) throws -> PollVoteDTO? {
+        PollVoteDTO.load(voteId: id, pollId: pollId, context: self)
+    }
+    
+    func removePollVote(with id: String, pollId: String) throws {
+        guard let dto = try pollVote(id: id, pollId: pollId) else {
+            // TODO: error
+            throw ClientError.Unknown()
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/PollVoteDTO.swift:192]: Highlighting the `TODO` e.g. `PollDoesNotExist(pollId:)`

File: Sources/StreamChat/Database/DatabaseSession.swift [hunk 8]
```diff
@@ -516,6 +562,64 @@ extension DatabaseSession {
...
+        if let vote = payload.vote {
+            if payload.eventType == .pollVoteRemoved {
+                if let dto = try? pollVote(id: vote.id, pollId: vote.pollId) {
+                    delete(pollVote: dto)
+                }
+            } else if payload.eventType == .pollVoteChanged {
+                var voteUpdated = false
+                let userId = vote.userId ?? "anon"
+                if let optionId = vote.optionId, !optionId.isEmpty {
+                    let id = "\(optionId)-\(vote.pollId)-\(userId)"
```
Comment (laevandus) [inline Sources/StreamChat/Database/DatabaseSession.swift:575]: This appeared already once before, wondering if it makes sense to put this into a separate `static func` for making sure we never have issues with `id` creation.

File: Sources/StreamChat/Models/ChatMessage.swift [hunk 9]
```diff
@@ -202,6 +202,8 @@ public struct ChatMessage {
     public var readByCount: Int { _readByCount }
 
     @CoreDataLazy internal var _readByCount: Int
+    
+    public let poll: Poll?
```
Comment (laevandus) [inline Sources/StreamChat/Models/ChatMessage.swift:206]: Since it is public and most of the public properties have documentation then I guess it would be nice to have something here as well although it is pretty self-explanatory.

File: Sources/StreamChat/Controllers/PollController/PollController.swift (hunk 1 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift:188]: It's actually `extraData` that we use currently, but we should use `custom` to align with other platforms. Now, the question is - do we want to add the new stuff the correct way, or we keep consistency with the rest of the codebase. Not sure, wdyt?

File: Sources/StreamChat/Controllers/PollController/PollController.swift (hunk 2 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift:57]: Nope, I will make them lazy

File: Sources/StreamChat/Database/DTOs/PollDTO.swift (hunk 4 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Database/DTOs/PollDTO.swift:23]: Wasn't able to make it optional, it defaults to 0 if it's an Int in Core Data. And we need this one to be optional.

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift (hunk 5 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Database/DTOs/PollVoteDTO.swift:99]: knew I forgot sth 😅 Yeah, new errors need to be introduced.

File: Sources/StreamChat/Database/DatabaseSession.swift (hunk 8 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Database/DatabaseSession.swift:575]: yes, good point, will add it

File: Sources/StreamChat/Models/Poll.swift [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    
+    /// The date and time when the poll was last updated.
+    /// This property is optional and may be `nil`.
+    public let updatedAt: Date?
+    
+    /// The count of votes received for the poll.
+    public let voteCount: Int
+    
+    /// A dictionary containing custom fields associated with the poll.
+    public let custom: [String: RawJSON]
```
Comment (nuno-vieira) [inline Sources/StreamChat/Models/Poll.swift:42]: Ideally this should be extraData, since we use it everywhere. IMO we should only rename it to custom in v5
```suggestion
    public let extraData: [String: RawJSON]
```

File: Sources/StreamChat/Models/Poll.swift [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    /// A boolean indicating whether the poll is closed.
+    public let isClosed: Bool
+    
+    /// The maximum number of votes allowed per user.
+    /// This property is optional and may be `nil`.
+    public let maxVotesAllowed: Int?
+    
+    /// A string representing the visibility of the voting process.
+    /// This property is optional and may be `nil`.
+    public let votingVisibility: String?
```
Comment (nuno-vieira) [inline Sources/StreamChat/Models/Poll.swift:57]: Don't we have a data structure to represent the `VotingVisiblity`? 🤔

File: Sources/StreamChat/Models/Poll.swift [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    
+    /// The user who created the poll.
+    /// This property is optional and may be `nil`.
+    public let createdBy: ChatUser?
```
Comment (nuno-vieira) [inline Sources/StreamChat/Models/Poll.swift:61]: How can this be optional? 🤔 Could the Poll be created by the system somehow?

File: Sources/StreamChat/Query/PollVoteListQuery.swift [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+import Foundation
+
+/// A query used for querying specific reactions from a message.
+public struct PollVoteListQuery: Encodable {
+    public var pollId: String
+    public var optionId: String?
+    /// The pagination information to query the reactions.
+    public var pagination: Pagination
+    /// The filter details to query the reactions.
```
Comment (nuno-vieira) [inline Sources/StreamChat/Query/PollVoteListQuery.swift:13]:
```suggestion
    /// The filter details to query the votes.
```

File: Sources/StreamChat/Query/PollVoteListQuery.swift [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+/// Non extra-data-specific filer keys for reaction list.
+public extension FilterKey where Scope: AnyVoteListFilterScope {
+    /// A filter key for matching the reaction type
+    static var optionId: FilterKey<Scope, String> { "option_id" }
+
+    /// A filter key for matching the user id of the reaction's author.
+    static var userId: FilterKey<Scope, UserId> { "user_id" }
+    
+    static var pollId: FilterKey<Scope, String> { "poll_id" }
+}
```
Comment (nuno-vieira) [inline Sources/StreamChat/Query/PollVoteListQuery.swift:55]: Copy pasting consequences here 😄 We should rename where Reaction is present

File: Sources/StreamChat/Repositories/PollsRepository.swift [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+    func createPoll(
+        name: String,
+        allowAnswers: Bool? = nil,
+        allowUserSuggestedOptions: Bool? = nil,
+        description: String? = nil,
+        enforceUniqueVote: Bool? = nil,
+        maxVotesAllowed: Int? = nil,
+        votingVisibility: String? = nil,
+        options: [PollOption]? = nil,
+        custom: [String: RawJSON]? = nil,
```
Comment (nuno-vieira) [inline Sources/StreamChat/Repositories/PollsRepository.swift:25]: nit: Since the repository/updater is internal, I find it more useful not to make these properties optional, so that when we use it in the controllers, we don't forget to reason about all the available properties

File: Sources/StreamChat/Repositories/PollsRepository.swift [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        } completion: { [weak self] error in
+            if error == nil {
+                self?.apiClient.request(
+                    endpoint: .removePollVote(
+                        messageId: messageId,
+                        pollId: pollId,
+                        voteId: voteId
+                    )
+                ) {
+                    completion?($0.error)
```
Comment (nuno-vieira) [inline Sources/StreamChat/Repositories/PollsRepository.swift:132]: If the request fails, shouldn't we add back the vote?

File: Sources/StreamChat/Repositories/PollsRepository.swift [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+            sort: sort,
+            filter: filter
+        )
+        apiClient.request(
+            endpoint: .queryPollVotes(pollId: pollId, queryPollVotesRequest: request),
+            completion: { [weak self] result in
+                guard let self else { return }
+                switch result {
+                case let .success(response):
+                    database.write { session in
```
Comment (nuno-vieira) [inline Sources/StreamChat/Repositories/PollsRepository.swift:224]: Fix compilation issue:
```suggestion
                    self.database.write { session in
```

File: Tests/StreamChatTests/Controllers/PollsControllers/PollController_Tests.swift [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+        let query = controller.ownVotesQuery
+        try client.databaseContainer.writeSynchronously { session in
+            try session.savePollVotes(payload: response, query: query, cache: nil)
+        }
+
+        XCTAssertEqual(controller.poll?.id, pollId)
+        XCTAssertEqual(controller.ownVotes.count, ownVotes.count)
+        
+        controller = nil
+        client = nil
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/Controllers/PollsControllers/PollController_Tests.swift:195]: Why is this here? This looks a bit suspicious 🤔

File: Tests/StreamChatTests/Controllers/PollsControllers/PollController_Tests.swift [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+        controller.castPollVote(answerText: nil, optionId: "123") {
+            completionError = $0
+        }
+
+        // Simulate network response with the error.
+        let networkError = TestError()
+        client.mockPollsRepository.castPollVote_completion?(networkError)
+
+        // Assert error is propagated.
+        AssertAsync.willBeEqual(completionError as? TestError, networkError)
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/Controllers/PollsControllers/PollController_Tests.swift:212]: nit: Whenever we can, it would be nice not to use AssertAsync, and just use regular expectations. We agreed on moving out from custom frameworks in tests.

File: Tests/StreamChatTests/Repositories/PollsRepository_Tests.swift [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+        let referenceEndpoint: Endpoint<PollPayloadResponse> = .createPoll(
+            createPollRequest: .init(name: pollName)
+        )
+        XCTAssertEqual(apiClient.request_endpoint, AnyEndpoint(referenceEndpoint))
+    }
+    
+    // MARK: - Closing poll
+    
+    func test_closePoll_whenSuccessfull() {
+        let completionCalled = expectation(description: "completion called")
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/Repositories/PollsRepository_Tests.swift:77]: Using expectations is much better and familiar to everyone 👍 It would be nice that we kept consistency and not use AssertAsync in someplaces and others expecations (unless it is necessary to use AssertAsync because of bad design)

**Review by nuno-vieira — COMMENTED** [review]
Looks great overall !👌  Added a couple of suggestions and questions 👍 

Btw, test_removingAllData() is failing because you need to provide there the new entities.

It would also be better if next time we split the PR maybe into the Poll List, and the Poll only. Otherwise it is a bit harder to revie the 4k lines of code, and Github does not handle well just a big PR 😄

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+        self.pollId = pollId
+        self.environment = environment
+        ownVotesQuery = PollVoteListQuery(
+            pollId: pollId,
+            optionId: nil,
+            pagination: .init(pageSize: 25, cursor: nil),
+            filter: .and(
+                [.equal(.userId, to: client.currentUserId ?? ""), .equal(.pollId, to: pollId)]
+            )
+        )
```
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift:134]: This query fetches the current user's votes. The backend allows max of 10 votes per user, but I put it to the max of the query just in case. Having this configurable shouldn't be possible for customers.

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+    
+    /// Casts a vote for a poll.
+    ///
+    /// - Parameters:
+    ///   - answerText: An optional text answer for the poll.
+    ///   - optionId: An optional identifier for the poll option.
+    ///   - completion: A closure to be called upon completion, with an optional `Error` if something went wrong.
+    public func castPollVote(
+        answerText: String?,
+        optionId: String?,
```
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift:161]: Good point, let me throw an error here

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+    }
+}
+
+/// Represents the visibility of votes in a poll.
+public enum VotingVisibility: String {
+    /// Votes are public and can be seen by everyone.
+    case `public`
+    /// Votes are anonymous and cannot be attributed to individual users.
+    case anonymous
+}
```
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift:243]: good point, will update just in case

File: Sources/StreamChat/Query/PollVoteListQuery.swift (hunk 13 above)
Comment (laevandus) [reply Sources/StreamChat/Query/PollVoteListQuery.swift:13]: One more above here

File: Sources/StreamChat/Models/Poll.swift (hunk 11 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Models/Poll.swift:57]: yep, forgot to change it, thx

File: Sources/StreamChat/Models/Poll.swift (hunk 12 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Models/Poll.swift:61]: it's optional in the spec. My assumption is for polls created anonymously or by the backend.

File: Sources/StreamChat/Query/PollVoteListQuery.swift (hunk 14 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Query/PollVoteListQuery.swift:55]: hahah 🤦 fixed them, thx

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 24]
```diff
@@ -715,37 +718,64 @@ public class ChatChannelController: DataController, DelegateCallable, DataStoreP
...
+    ///   - votingVisibility: The visibility setting for the poll.
+    ///   - options: Optional array of predefined poll options.
+    ///   - custom: Optional dictionary for any custom data associated with the poll.
+    ///   - completion: A closure to be executed once the poll is created, returning either a `MessageId` on success or an `Error` on failure.
+    public func createPoll(
+        name: String,
+        allowAnswers: Bool? = nil,
+        allowUserSuggestedOptions: Bool? = nil,
+        description: String? = nil,
+        enforceUniqueVote: Bool? = nil,
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:755]: The problem with this is that it is not explicit, so the default value is not transparent to the developer. We at least need to say it in the code comments. 

For example, in `ThreadListQuery` all the values are also nullable. But if I did not put the values from the backend on the SDK, then the developer would need to go to the Backend API to check what the defaults

[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:755]: I documented the defaults in the method - safer to pass nil over a default value.

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+
+extension PollController {
+    struct Environment {
+        var pollObserverBuilder: (
+            _ isBackgroundMappingEnabled: Bool,
+            _ database: DatabaseContainer,
+            _ fetchRequest: NSFetchRequest<PollDTO>,
+            _ itemCreator: @escaping (PollDTO) throws -> Poll,
+            _ fetchedResultsControllerType: NSFetchedResultsController<PollDTO>.Type
+        ) -> EntityDatabaseObserverWrapper<Poll, PollDTO> = {
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollController.swift:267]: Shorter `= EntityDatabaseObserverWrapper.init` should also work here but so important. Same with the other observer.

File: Sources/StreamChat/Controllers/PollController/PollVoteListController.swift [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+            _ apiClient: APIClient
+        ) -> PollsRepository = PollsRepository.init
+
+        var createPollListDatabaseObserver: (
+            _ isBackgroundMappingEnabled: Bool,
+            _ database: DatabaseContainer,
+            _ fetchRequest: NSFetchRequest<PollVoteDTO>,
+            _ itemCreator: @escaping (PollVoteDTO) throws -> PollVote
+        )
+            -> ListDatabaseObserverWrapper<PollVote, PollVoteDTO> = {
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollVoteListController.swift:189]: Alternative: `=  ListDatabaseObserverWrapper.init` (not important)

**Comment by laevandus** [comment]
I guess state layer support comes with a separate PR?

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 27]
```diff
@@ -715,37 +718,64 @@ public class ChatChannelController: DataController, DelegateCallable, DataStoreP
...
+            votingVisibility: votingVisibility?.rawValue,
+            options: options,
+            custom: extraData
+        ) { [weak self] result in
+            guard let self else { return }
+            switch result {
+            case let .success(poll):
+                self.createNewMessage(text: "", poll: poll, completion: completion)
+            case let .failure(error):
+                completion(.failure(error))
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:778]: We should use `callback {}` here for calling the completion handler like controllers do.

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+        }
+        
+        pollsRepository.castPollVote(
+            messageId: messageId,
+            pollId: pollId,
+            answerText: answerText,
+            optionId: optionId,
+            currentUserId: client.currentUserId,
+            query: ownVotesQuery,
+            completion: completion
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollController.swift:179]: We should use `callback {}` here for completion handler.

File: Sources/StreamChat/Controllers/PollController/PollController.swift [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+    ///   - completion: A closure to be called upon completion, with an optional `Error` if something went wrong.
+    public func removePollVote(
+        voteId: String,
+        completion: ((Error?) -> Void)? = nil
+    ) {
+        pollsRepository.removePollVote(
+            messageId: messageId,
+            pollId: pollId,
+            voteId: voteId,
+            completion: completion
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/PollController/PollController.swift:196]: We should use `callback {}` here for completion handler.

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 24 above)
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:755]: IMO, I think it is actually safer to have the default values on the SDK side. Because if the customer does not pass anything, and if the backend changes the values (hopefully this does not happen) it could be a silent breaking change for the customer.

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift [hunk 30]
```diff
@@ -46,7 +46,14 @@ class PollVoteDTO: NSManagedObject {
...
     static func load(voteId: String, pollId: String, context: NSManagedObjectContext) -> PollVoteDTO? {
         let request = fetchRequest(for: voteId, pollId: pollId)
-        return load(by: request, context: context).first
+        let results = load(by: request, context: context)
+        if results.count > 1 {
+            for i in 1..<results.count {
+                let dto = results[i]
+                context.delete(dto)
+            }
+        }
```
Comment (nuno-vieira) [inline Sources/StreamChat/Database/DTOs/PollVoteDTO.swift:55]: Can you explain this one? Maybe add a comment? I get the feeling this should not be part of the `load` function, since it is doing more than just loading, should this logic be outside of the load function?

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 24 above)
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:755]: I disagree with that - if the backend says a value is optional, then we shouldn't be making assumptions by sending default values. We already had few issues because of this, last example was in the video SDK last week.
As for your concern - the no-breaking changes policy applies for both frontend and backend, so as long as we respect the contract it's gonna be ok.

File: Sources/StreamChat/Database/DTOs/PollVoteDTO.swift (hunk 30 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Database/DTOs/PollVoteDTO.swift]: actually removed it, found a better way

File: Sources/StreamChat/Controllers/PollController/PollController.swift (hunk 28 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/PollController/PollController.swift]: fixed

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 27 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:778]: fixed

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 24 above)
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:755]: It depends on the use case ofc. In some scenarios, making the default values explicit on the SDK is easier for the customer, because it does not require the customer to read the API documentation. Even if we add a comment saying what is the default, it can easily be out of sync with the backend. For example, queries, it is useful to see the default values in the SDK

File: Sources/StreamChat/Database/DTOs/PollDTO.swift [hunk 31]
```diff
@@ -0,0 +1,180 @@
...
+        pollDto.voteCount = payload.voteCount
+        pollDto.custom = try JSONEncoder.default.encode(payload.custom)
+        pollDto.voteCountsByOption = payload.voteCountsByOption
+        pollDto.isClosed = payload.isClosed ?? false
+        if let maxVotesAllowed = payload.maxVotesAllowed {
+            pollDto.maxVotesAllowed = NSNumber(value: maxVotesAllowed)
+        }
+        pollDto.votingVisibility = payload.votingVisibility
+        
+        pollDto.createdBy = UserDTO.loadOrCreate(id: payload.createdById, context: self, cache: cache)
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/PollDTO.swift:125]: Seems like we missing a line here which stores the user. `payload.createdBy` is `UserPayload` and we are not saving it here so this can return nil.

**Review by nuno-vieira — APPROVED** [review]

**Review by laevandus — APPROVED** [review]
